### PR TITLE
Initial test implementation of ADC helper functions

### PR DESCRIPTION
**Pull request by HarkonenBade**
Test implementation of the helper functions for various ADC internal sensors. Currently untested on actual hardware.

### CODE REVIEW

File: src/adc.rs [hunk 1]
```diff
@@ -46,9 +58,10 @@ pub struct Adc {
     sample_time: AdcSampleTime,
     align: AdcAlign,
     precision: AdcPrecision,
+    stored_cfg: Option<(AdcSampleTime, AdcAlign, AdcPrecision)>,
```
Comment (david-sawatzke) [inline src/adc.rs]: Couldn't you just store this on the stack for the functions, e.g. the stash (should then be renamed) just gives the tuple back and restore just takes it. There's no real need to always allocate space for this (i think)

Comment (HarkonenBade) [reply src/adc.rs]: This could be done, I suspect the storage use is incredibly minimal but I can correct.

File: src/adc.rs [hunk 2]
```diff
@@ -241,6 +254,52 @@ impl VTemp {
...
+            vtemp.enable(adc);
+
+            // Double read of vdda to allow sufficient startup time for the temp sensor
+            adc.read_vdda();
+        }
+        let vdda = adc.read_vdda();
+
+        adc.stash_cfg();
+
+        adc.set_align(AdcAlign::Right);
```
Comment (david-sawatzke) [inline src/adc.rs]: Do you really need to duplicate this code? Couldn't you just put this in a non-public `get_raw` method (or something)?

File: src/adc.rs [hunk 3]
```diff
@@ -241,6 +254,52 @@ impl VTemp {
...
+    }
+
+    pub fn read_vtemp(adc: &mut Adc) -> i16 {
+        let mut vtemp = Self::new();
+        let vtemp_preenable = vtemp.is_enabled(&adc);
+
+        if !vtemp_preenable {
+            vtemp.enable(adc);
+
+            // Double read of vdda to allow sufficient startup time for the temp sensor
```
Comment (david-sawatzke) [inline src/adc.rs]: Wouldn't using a delay reference be nicer here?

File: src/adc.rs (hunk 2 above)
Comment (HarkonenBade) [reply src/adc.rs]: I could put in a reset to default settings function.

File: src/adc.rs (hunk 3 above)
Comment (HarkonenBade) [reply src/adc.rs]: Yes, though I was trying to avoid forcing the user to provide a delay if possible.

File: src/adc.rs (hunk 1 above)
Comment (HarkonenBade) [reply src/adc.rs]: It would be saving 3 bytes on a structure that can only ever be instantiated once.

Comment (HarkonenBade) [reply src/adc.rs]: Done

File: src/adc.rs (hunk 3 above)
Comment (HarkonenBade) [reply src/adc.rs]: Made an option

File: src/adc.rs (hunk 2 above)
Comment (HarkonenBade) [reply src/adc.rs]: Done

**Comment by HarkonenBade** [comment]
Are there any more structural features that people think should change? Otherwise I will start working on documenting and stuff to make this into a finished PR for merging.

**Comment by HarkonenBade** [comment]
@therealprof @david-sawatzke Ok, should be in final shape now. Would appreciate any final comments or requests, otherwise should be good to merge.

File: Cargo.toml [hunk 4]
```diff
@@ -42,6 +42,7 @@ version = "0.2.2"
 
 [dev-dependencies]
 panic-halt = "0.2.0"
+cortex-m-semihosting = "0.3.1"
```
Comment (therealprof) [inline Cargo.toml]:
```suggestion
cortex-m-semihosting = "0.3.1"
```

Seems superfluous, no?

Comment (HarkonenBade) [reply Cargo.toml]: Apologise, it was from my previous semihosting based example

**Comment by HarkonenBade** [comment]
This is what I get for trying to be fast and commit code directly on ghub :/

**Pull request merged**